### PR TITLE
documentation: add documentation for XGBoost

### PR DESCRIPTION
**Pull request by eslesar-aws**
*Description of changes:*

Added rst files to generate XGBoost API reference content and explain usage of open source XGBoost in SageMaker.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have used the regional endpoint when creating S3 and/or STS clients (if appropriate)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: doc/using_xgboost.rst [hunk 1]
```diff
@@ -0,0 +1,200 @@
+###########################################
+Using XGBoost with the SageMaker Python SDK
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: should this also be "Use"?

File: doc/using_xgboost.rst [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+It implements a technique known as gradient boosting on trees, which performs remarkably well in machine learning competitions.
+
+Amazon SageMaker supports two ways to use the XGBoost algorithm:
+
+ * XGBoost built-in algorithm
+ * XGBoost open source algorithm
+
+The XGBoost open source algorithm provides the following benefits over the built-in algorithm:
+
+* Latest version - The open source XGBoost algorithm supports XGBoost verson 1.0 that has better performance scaling on multi-core instances and
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: * verson --> version
* (personal preference) "version 1.0 that" --> "version 1.0, which"

File: doc/using_xgboost.rst [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  improved stability for distributed training.
+* Flexibility - Because you write a custom training script, the open source XGBoost algorithm, you can add custom pre- and post-processing logic,
+  run additional code after training, and take advantage of the full range of XGBoost functions. For example, cross-validation support.
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: I think I would flip this around: "Take advantage of the full range of XGBoost functionality, such as cross-validation support. You can add custom pre- and post-processing logic and run additional code after training."

File: doc/using_xgboost.rst [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+* Scalability - The XGBoost open source algorithm has a more efficient implementation of distributed training,
+  which enables it to scale out to more instances and reduce out-of-memory errors.
+* Exensibility - Because the open source XGBoost container is open source,
+  you can extend the container to install additional libraries and change the version of XGBoost that the container uses.
+  For more information, see `SageMaker XGBoost Container <https://github.com/aws/sagemaker-xgboost-container>`__.
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: is there a notebook that shows extending the image? if so, I think that would be a better link because the README of the repo launches into how to build and test the image, which isn't particularly relevant for people who just want to use it as a base image in their Dockerfile

File: doc/using_xgboost.rst [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+1. `Prepare a training script <#prepare-a-training-script>`_
+2. |create xgboost estimator|_
+3. |call fit|_
+
+Prepare a Training Script
+-------------------------
+
+A typical training script loads data from the input channels, configures training with hyperparameters, trains a model,
+and saves a model to model_dir so that it can be hosted later.
+Hyperparameters are passed to your script as arguments and can be retrieved with an argparse.ArgumentParser instance.
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: monospace for `argparse.ArgumentParser`. also might be helpful to link to https://docs.python.org/3/library/argparse.html

File: doc/using_xgboost.rst [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    if __name__ == '__main__':
+        parser = argparse.ArgumentParser()
+
+    # Hyperparameters are described here
+    parser.add_argument('--num_round', type=int)
+    parser.add_argument('--max_depth', type=int, default=5)
+    parser.add_argument('--eta', type=float, default=0.2)
+    parser.add_argument('--objective', type=str, default='reg:squarederror')
+    
+    # Sagemaker specific arguments. Defaults are set in the environment variables.
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: Sagemaker --> SageMaker

File: doc/using_xgboost.rst [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+A typical training script loads data from the input channels, configures training with hyperparameters, trains a model,
+and saves a model to model_dir so that it can be hosted later.
+Hyperparameters are passed to your script as arguments and can be retrieved with an argparse.ArgumentParser instance.
+
+For a complete example of an XGBoost training script, see https://github.com/awslabs/amazon-sagemaker-examples/blob/master/introduction_to_amazon_algorithms/xgboost_abalone/abalone.py.
+
+Let's look at the main elements of the script. Starting with the ``__main__`` guard,
+use a parser to read the hyperparameters passed to the estimator when creating the training job.
+These hyperparameters are made available as arguments to our input script.
+We also parse a number of Amazon SageMaker-specific environment variables to get information about the training environment,
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: I think you could do something similar as with the other frameworks, where some of the common variables are listed, and then there's a link to the exhaustive list, e.g. https://sagemaker.readthedocs.io/en/stable/using_mxnet.html#for-versions-1-3-and-higher

File: doc/using_xgboost.rst [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    model_location = args.model_dir + '/xgboost-model'
+    pkl.dump(bst, open(model_location, 'wb'))
+    logging.info("Stored trained model at {}".format(model_location))
+
+In the training script, you can customize the inference behavior by implementing the follwing functions:
+* ``input_fn`` - how input data is handled.
+* ``predict_fn`` - how the model is invokedfunction, and how the response is returned ).
+* ``output_fn`` - How the response data is handled
+
+These functions are optional. If you want to use the default implementations, do not implement them in your training script.
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: I think it's a little strange to talk about inference functions in a _training_ script - maybe move this to the "Deploy" section

File: doc/using_xgboost.rst [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        image_name=container,
+        role=role, 
+        train_instance_count=1,
+        train_instance_type="ml.m5.2xlarge",
+        framework_version="0.90-1",
+        output_path="s3://{}/{}/{}/output".format(bucket, prefix, "xgboost-script-mode"),
+        train_use_spot_instances=train_use_spot_instances,
+        train_max_run=train_max_run,
+        train_max_wait=train_max_wait,
+        checkpoint_s3_uri=checkpoint_s3_uri
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: do we need all of these? since this is just a basic example, I think only `entry_point`, `hyperparameters`, `role`, `train_instance_count`, `train_instance_type`, and `framework_version` are necessary

File: doc/using_xgboost.rst [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        dtrain=dtrain,
+        evals=watchlist,
+        num_boost_round=(args.num_round - n_iterations_prev_run),
+        xgb_model=prev_checkpoint,
+        callbacks=callbacks
+    )
+    
+    model_location = args.model_dir + '/xgboost-model'
+    pkl.dump(bst, open(model_location, 'wb'))
+    logging.info("Stored trained model at {}".format(model_location))
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: might be worth calling out separately that the script needs to save the model and where it has to be saved

File: doc/xgboost.rst [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+-------
+
+The Amazon SageMaker XGBoost open source framework algorithm.
+
+.. autoclass:: sagemaker.xgboost.estimator.XGBoost
+    :members:
+    :undoc-members:
+    :show-inheritance:
+    :inherited-members:
+    :exclude-members: image, num_factors, predictor_type, epochs, clip_gradient, mini_batch_size, feature_dim, eps, rescale_grad, bias_lr, linear_lr, factors_lr, bias_wd, linear_wd, factors_wd, bias_init_method, bias_init_scale, bias_init_sigma, bias_init_value, linear_init_method, linear_init_scale, linear_init_sigma, linear_init_value, factors_init_method, factors_init_scale, factors_init_sigma, factors_init_value
```
Comment (laurenyu) [inline doc/xgboost.rst]: you don't need this for the framework estimator

File: doc/xgboost.rst [hunk 12]
```diff
@@ -0,0 +1,22 @@
+XGBoost
```
Comment (laurenyu) [inline doc/xgboost.rst]: maybe the title should also include that this is the open source version?

File: doc/using_xgboost.rst (hunk 4 above)
Comment (eslesar-aws) [reply doc/using_xgboost.rst]: I can't find any such notebook. I get that this link isn't ideal, but do you think nothing would be better?

File: doc/using_xgboost.rst (hunk 10 above)
Comment (eslesar-aws) [reply doc/using_xgboost.rst]: Meant to ask about that in email. So is the extra /xgboost-model subdir within model_dir necessary here? Or can it be saved anywhere within model_dir?

File: doc/using_xgboost.rst (hunk 8 above)
Comment (eslesar-aws) [reply doc/using_xgboost.rst]: I can move it to the Deploy section, but I've talked about a "training script". Should we call it something else, or suggest that you could put it in a separate "inference script"?

File: doc/using_xgboost.rst [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+
+.. contents::
+
+eXtreme Gradient Boosting (XGBoost) is a popular and efficient machine learning algorithm used for regression and classification tasks on tabular datasets.
+It implements a technique known as gradient boosting on trees, which performs remarkably well in machine learning competitions.
+
+Amazon SageMaker supports two ways to use the XGBoost algorithm:
+
+ * XGBoost built-in algorithm
+ * XGBoost open source algorithm
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: maybe "open source library"? (taking inspiration from the first sentence of the [XGBoost docs](https://xgboost.readthedocs.io/en/latest/))

File: doc/using_xgboost.rst [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+The XGBoost open source algorithm provides the following benefits over the built-in algorithm:
+
+* Latest version - The open source XGBoost algorithm supports XGBoost version 1.0, which has better performance scaling on multi-core instances and
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: might be better to remove the version here and just link to https://github.com/aws/sagemaker-python-sdk/tree/master/src/sagemaker/xgboost#xgboost-sagemaker-estimators-and-models, so that fewer places have to be updated with each version upgrade.

although it seems like that page has already gone out of date...

File: doc/using_xgboost.rst (hunk 4 above)
Comment (laurenyu) [reply doc/using_xgboost.rst]: is there documentation for finding the image URI? that's probably the most relevant thing for customers looking to extend the image.

my hesitancy with linking to the instructions for building the image from scratch is that we've had previous GitHub issues where people thought they needed to build the image because that's the README they found, and they hadn't realized that they could just use the pre-built version. to be fair, that might just mean we need to overhaul the framework repository READMEs...

File: doc/using_xgboost.rst [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+        dtrain=dtrain,
+        evals=watchlist,
+        num_boost_round=(args.num_round - n_iterations_prev_run),
+        xgb_model=prev_checkpoint,
+        callbacks=callbacks
+    )
+    
+    model_location = args.model_dir + '/xgboost-model'
+    pkl.dump(bst, open(model_location, 'wb'))
+    logging.info("Stored trained model at {}".format(model_location))
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: all of this should be indented to match l. 95

File: doc/using_xgboost.rst (hunk 10 above)
Comment (laurenyu) [reply doc/using_xgboost.rst]: bumped on the email thread - I honestly don't know in this case

File: doc/using_xgboost.rst [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+
+Create an Estimator
+-------------------
+After you create your training script, create an instance of the :class:`sagemaker.xgboost.XGBoost` estimator.
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: I don't think `:class:`sagemaker.xgboost.XGBoost` links to anything automatically here

File: doc/using_xgboost.rst [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+Pass an IAM role that has the permissions necessary to run an Amazon SageMaker training job,
+the type and number of instances to use for the training job,
+and a dictionary of the hyperparameters to pass to the training script.
+
+.. code::
+
+    from sagemaker.session import s3_input
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: you can delete the `s3_input` import

File: doc/using_xgboost.rst [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    from sagemaker.xgboost.estimator import XGBoost
+
+    xgb_script_mode_estimator = XGBoost(
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: I'd just call the variable `xgb` or `xgb_estimator`

File: doc/using_xgboost.rst [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+.. code::
+
+    predictor = xgb_script_mode_estimator.deploy(initial_instance_count=1, instance_type="ml.m5.xlarge")
+    test_data = xgboost.DMatrix('/path/to/data')
+    predictor.predict(test_data)
+
+Customize inference
+-------------------
+
+In the script that you provide, you can customize the inference behavior by implementing the follwing functions:
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: follwing --> following

File: doc/using_xgboost.rst [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+* ``input_fn`` - how input data is handled.
+* ``predict_fn`` - how the model is invokedfunction, and how the response is returned ).
```
Comment (laurenyu) [inline doc/using_xgboost.rst]: messed up copy/paste?

File: doc/using_xgboost.rst (hunk 8 above)
Comment (laurenyu) [reply doc/using_xgboost.rst]: I think I might go with something like "in your inference script, which can be in the same file as your training script from above"

File: doc/using_xgboost.rst (hunk 4 above)
Comment (eslesar-aws) [reply doc/using_xgboost.rst]: I can't find anything other than the use of the get_image_uri function itself in code examples. 

I changed the link to point to the example notebook that extends the pytorch container (https://github.com/awslabs/amazon-sagemaker-examples/blob/master/advanced_functionality/pytorch_extending_our_containers/pytorch_extending_our_containers.ipynb). That is the only example that I can find on extending containers.

File: doc/using_xgboost.rst (hunk 13 above)
Comment (eslesar-aws) [reply doc/using_xgboost.rst]: I'm going to punt on this for now. I don't think it will confuse anyone as is.

File: doc/using_xgboost.rst (hunk 10 above)
Comment (eslesar-aws) [reply doc/using_xgboost.rst]: The intro section says you have to save the model to model_dir, and I added a comment to the part of the script where it saves the model (and an add_argument line for SM_MODEL_DIR). I'll add a section for save model when I get more specific information.

**Comment by laurenyu** [comment]
error from the Sphinx check:

> Warning, treated as error:
/codebuild/output/src225721869/src/github.com/aws/sagemaker-python-sdk/.tox/sphinx/lib/python3.6/site-packages/sagemaker/xgboost/estimator.py:docstring of sagemaker.xgboost.estimator.XGBoost:11:Unexpected indentation.

**Comment by eslesar-aws** [comment]
I don't see in the xgboost estimator.py file where this indentation is. I didn't change this file.

**Comment by laurenyu** [comment]
I think it's because you added `doc/xgboost.rst`, which includes the estimator docstrings. My guess is it's these two lines: https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/xgboost/estimator.py#L70-L71

**Comment by laurenyu** [comment]
> Warning, treated as error:
> /codebuild/output/src121143675/src/github.com/aws/sagemaker-python-sdk/.tox/sphinx/lib/python3.6/site-packages/sagemaker/xgboost/model.py:docstring of sagemaker.xgboost.model.XGBoostModel.prepare_container_def:9:Unexpected indentation.

I think this is https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/xgboost/model.py#L112. Also not entirely sure about the indentation at https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/xgboost/model.py#L119

File: src/sagemaker/xgboost/model.py [hunk 21]
```diff
@@ -108,16 +108,15 @@ def __init__(
         self.model_server_workers = model_server_workers
 
     def prepare_container_def(self, instance_type, accelerator_type=None):
-        """Return a container definition with framework configuration set in model environment
-            variables.
+        """Return a container definition with framework configuration set in model environment variables.
```
Comment (laurenyu) [inline src/sagemaker/xgboost/model.py:111]: from the failed build:

```
[7;33m************* Module sagemaker.xgboost.model[0m
src/sagemaker/xgboost/model.py:111:0: C0301: [1mLine too long (105/100)[0m ([1mline-too-long[0m)
```

can you split this line up? (and the second line should be indented at the same level as `"""`)
